core/drivers/crypto/caam/cipher/caam_cipher_mac.c

```diff
@@ -0,0 +1,861 @@
+	bool realloc = false;
+
+	CIPHER_TRACE("Length=%zu - %s", dupdate->src.length,
+		     dupdate->encrypt ? "Encrypt" : " Decrypt");
+
+	if (dupdate->src.length) {
+		srcbuf.data = dupdate->src.data;
+		srcbuf.length = dupdate->src.length;
+		srcbuf.paddr = virt_to_phys(dupdate->src.data);
+
+		if (!srcbuf.paddr) {
+			CIPHER_TRACE("Bad Src address");
+			return TEE_ERROR_GENERIC;
+		}
+
+		if (!caam_mem_is_cached_buf(dupdate->src.data,
+					    dupdate->src.length))
+			srcbuf.nocache = 1;
+	}
+
+	if (dupdate->last) {
+		retstatus = caam_set_or_alloc_align_buf(dupdate->dst.data,
+							&dst_align,
+							dupdate->dst.length,
+							&realloc);
+		if (retstatus != CAAM_NO_ERROR) {
+			CIPHER_TRACE("Destination buffer reallocation error");
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+	}
+
+	/* Calculate the total data to be handled */
+	fullSize = ctx->blockbuf.filled + srcbuf.length;
+	if (!dupdate->last) {
+		if (fullSize < ctx->alg->size_block) {
+			size_topost = srcbuf.length;
+		} else {
+			size_topost = fullSize % ctx->alg->size_block;
+
+			/*
+			 * In case there is no data to save and because it's
+			 * not the final operation, ensure that a block of data
+			 * is kept for the final operation.
+			 */
+			if (!size_topost)
+				size_topost = ctx->alg->size_block;
+
+			/* Total size that is a cipher block multiple */
+			size_todo = fullSize - size_topost;
+		}
+	}
+
+	CIPHER_TRACE("FullSize %zu - posted %zu - todo %zu", fullSize,
+		     size_topost, size_todo);
+
+	if (size_todo || dupdate->last) {
+		size_inmade = srcbuf.length - size_topost;
+		srcbuf.length = size_inmade;
+
+		ret = run_cmac_desc(ctx, &srcbuf, &dst_align, dupdate->last);
+
+		srcbuf.length = dupdate->src.length;
+
+		if (ret == TEE_SUCCESS && dupdate->last && realloc)
+			memcpy(dupdate->dst.data, dst_align.data,
+			       dupdate->dst.length);
+	} else {
+		ret = TEE_SUCCESS;
+	}
+
+	if (size_topost) {
+		CIPHER_TRACE("Post %zu of input len %zu made %zu", size_topost,
+			     srcbuf.length, size_inmade);
+		if (caam_cpy_block_src(&ctx->blockbuf, &srcbuf, size_inmade) !=
+		    CAAM_NO_ERROR)
+			ret = TEE_ERROR_GENERIC;
+	}
+
+out:
+	if (realloc)
+		caam_free_buf(&dst_align);
+
+	return ret;
+}
+
+/*
+ * Initialization of the cipher MAC operation.
+ *
+ * @ctx  Operation software context
+ * @key  Input key to compute
+ * @len  Key length
+ */
+static TEE_Result do_cmac_init(struct crypto_mac_ctx *ctx, const uint8_t *key,
+			       size_t len)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t *iv_tmp = NULL;
+	struct drvcrypt_cipher_init dinit = { };
+	struct crypto_mac *mac = to_mac_ctx(ctx);
+	struct cipherdata *macdata = mac->ctx;
+
+	if (macdata->mode != TEE_CHAIN_MODE_CMAC) {
+		/* Allocate temporary IV initialize with 0's */
+		iv_tmp = caam_calloc(macdata->alg->size_ctx);
+		if (!iv_tmp)
+			return TEE_ERROR_OUT_OF_MEMORY;
+	} else {
+		/*
+		 * Check if the context registers is allocated to free it,
```

Comment:
`context registers are` or `context register is`
same at lines 621/621.

Reply:
Fixed